File: haskell/complex-numbers/src/ComplexNumbers.hs
module ComplexNumbers
(Complex,
 conjugate,
 abs,
 real,
 imaginary,
 mul,
 add,
 sub,
 div,
 complex) where

import Prelude hiding (div, abs)

-- Data definition -------------------------------------------------------------
data Complex a = Complex a a deriving(Eq, Show)

complex :: (a, a) -> Complex a
complex (r, i) = Complex r i

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate (Complex r i) = Complex r (-i)

abs :: Floating a => Complex a -> a
abs (Complex r i) = sqrt (r * r + i * i)

real :: Num a => Complex a -> a
real (Complex r _) = r

imaginary :: Num a => Complex a -> a
imaginary (Complex _ i) = i

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (Complex r1 i1) (Complex r2 i2) = Complex (r1 * r2 - i1 * i2) (i1 * r2 + r1 * i2)

add :: Num a => Complex a -> Complex a -> Complex a
add (Complex r1 i1) (Complex r2 i2) = Complex (r1 + r2) (i1 + i2)

sub :: Num a => Complex a -> Complex a -> Complex a
sub (Complex r1 i1) (Complex r2 i2) = Complex (r1 - r2) (i1 - i2)

div :: Fractional a => Complex a -> Complex a -> Complex a
div (Complex r1 i1) (Complex r2 i2) = Complex r i
    where
        denominator = r2 * r2 + i2 * i2
        r = (r1 * r2 + i1 * i2) / denominator
        i = (i1 * r2 - r1 * i2) / denominator

File: prolog/complex-numbers/complex_numbers.pl
real((R,_), R).
imaginary((_,I), I).

conjugate((R,I), (R, Ix)) :- Ix is 0-I.
abs((R, I), A) :- A is sqrt(R*R + I*I).

add((R1,I1),(R2, I2),(R3, I3)) :- R3 is R1 + R2, I3 is I1 + I2.
sub((R1,I1),(R2, I2),(R3, I3)) :- R3 is R1 - R2, I3 is I1 - I2.

mul((R1,I1),(R2, I2),(R3, I3)) :- R3 is R1 * R2 - I1 * I2, I3 is I1 * R2 + R1 * I2.
div((R1,I1),(R2, I2),(R3, I3)) :- 
    conjugate((R2, I2), C2),
    mul((R1,I1), C2, (NR,NI)),
    mul((R2,I2), C2, (DR,_)),
    R3 is NR / DR,
    I3 is NI / DR.
